### PR TITLE
Feature/init project and smart contract

### PR DESCRIPTION
**Pull request by SeddikBellamine**
Init Hardhat project
implement whitelist smart contract
implement unit tests for the whitelist smart contract
configure drone ci of default pipeline and deployment pipelines

### CODE REVIEW

File: README.md
Comment (Le-Caignec) [inline README.md:1]: Add details on the readme about the repo and its purpose

File: .eslintrc.json [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+    "airbnb-typescript/base",
+    "plugin:@typescript-eslint/recommended",
+    "prettier"
+  ],
+  "parser": "@typescript-eslint/parser",
+  "parserOptions": {
+    "ecmaVersion": "latest",
+  },
+  "rules": {}
+}
```
Comment (Le-Caignec) [inline .eslintrc.json]:
```suggestion
{
  "plugins": [
    "@typescript-eslint", // mandatory typescript plugin
    "sonarjs" // sonar plugin for code analysis
  ],
  "extends": [
    "airbnb-base", // use instead on eslint recommended as a stricter alternative
    "airbnb-typescript/base", // airbnb rules overrides for typescript
    "plugin:@typescript-eslint/recommended",
    "plugin:sonarjs/recommended", // sonar
    "prettier" // compatibility with prettier (disable conflicting rules) NB: this is the `eslint-config-prettier` it should not be duplicated
  ],
  "parser": "@typescript-eslint/parser",
  "parserOptions": {
    "ecmaVersion": "latest",
    "project": "./tsconfig.json" // Specify the path to your tsconfig.json file
  },
  "rules": {}
}
```
use .eslint propose by hardhat. Moreover the last config didn't work on my side

File: package.json [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    "mocha": "^10.2.0",
+    "prettier": "^3.0.3",
+    "prettier-plugin-solidity": "^1.1.3",
+    "ts-node": "^10.9.1",
+    "typescript": "^5.2.2"
+  },
+  "dependencies": {
+    "@nomicfoundation/hardhat-chai-matchers": "^2.0.2",
+    "@openzeppelin/contracts": "^4.9.3"
+  }
```
Comment (Le-Caignec) [inline package.json]:
```suggestion
  "devDependencies": {
    "@nomicfoundation/hardhat-toolbox": "^1.0.2",
    "@nomiclabs/hardhat-etherscan": "^3.1.7",
    "hardhat": "^2.14.1",
    "prettier": "^2.8.8",
    "prettier-plugin-solidity": "^1.1.3"
  },
  "dependencies": {
    "@openzeppelin/contracts": "^4.8.2"
  }
```
clean useless dependency

File: package.json [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    "removeRessourceFromWhitelist": "hardhat run scripts/removeRessourceFromWhitelist.ts --network bellecour",
+    "transferOwnership": "hardhat run scripts/transferOwnership.ts --network bellecour",
+    "verify": "hardhat verify",
+    "clean": "hardhat clean",
+    "lint": "eslint **/*.ts",
+    "lint:fix": "eslint **/*.ts --fix",
+    "typecheck": "tsc --noEmit",
+    "format": "prettier --plugin=./node_modules/prettier-plugin-solidity/src/index.js --write \"{scripts,test,contracts}/**/*.{ts,sol}\"",
+    "check-format": "prettier --plugin=./node_modules/prettier-plugin-solidity/src/index.js --check \"{scripts,test,contracts}/**/*.{ts,sol}\""
+  },
```
Comment (Le-Caignec) [inline package.json]:
```suggestion
  "scripts": {
    "compile": "hardhat compile",
    "test": "hardhat test",
    "coverage": "hardhat coverage",
    "deploy": "hardhat run scripts/deploy.ts --network bellecour",
    "addRessourceToWhitelist": "hardhat run scripts/addRessourceToWhitelist.ts --network bellecour",
    "removeRessourceFromWhitelist": "hardhat run scripts/removeRessourceFromWhitelist.ts --network bellecour",
    "transferOwnership": "hardhat run scripts/transferOwnership.ts --network bellecour",
    "verify": "hardhat verify",
    "clean": "hardhat clean",
    "lint": "eslint . --ext .ts",
    "format": "prettier --plugin=./node_modules/prettier-plugin-solidity/src/index.js --write \"{scripts,test,contracts}/**/*.{ts,sol}\"",
    "check-format": "prettier --plugin=./node_modules/prettier-plugin-solidity/src/index.js --check \"{scripts,test,contracts}/**/*.{ts,sol}\""
  },
```
remove useless scripts (lint:fix + typecheck)

File: hardhat.config.ts [hunk 4]
```diff
@@ -0,0 +1,58 @@
+import '@nomicfoundation/hardhat-toolbox';
+import 'hardhat-deploy';
+import { HardhatUserConfig } from 'hardhat/config';
+const { WALLET_PRIVATE_KEY } = process.env;
```
Comment (Le-Caignec) [inline hardhat.config.ts]:
```suggestion
import { HardhatUserConfig } from 'hardhat/config';
import '@nomicfoundation/hardhat-toolbox';
import '@nomiclabs/hardhat-etherscan';

const { WALLET_PRIVATE_KEY } = process.env;
```
remove useless library for deploying SC. It's not currently the best method to do it

File: utils/utils.ts [hunk 5]
```diff
@@ -0,0 +1,10 @@
+import fs from 'fs/promises';
+
+// if you change the name of the file, you need to change the name in the pipeline too
+const SMART_CONTRACT_ADDRESS_FILE = '.smart-contract-address';
```
Comment (Le-Caignec) [inline utils/utils.ts]: move that in a config.ts file

File: scripts/deploy.ts [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+};
+// We recommend this pattern to be able to use async/await everywhere
+// and properly handle errors.
+func(hre).catch((error) => {
+  console.error(error);
+  process.exitCode = 1;
+});
+
+export default func;
+func.tags = ['IExecWhitelist'];
```
Comment (Le-Caignec) [inline scripts/deploy.ts]:
```suggestion
import { ethers } from 'hardhat';
import { saveSmartContractAddress } from '../utils/utils';

async function main() {
  console.log('Starting deployment...');
  const [deployer] = await ethers.getSigners();
  console.log('Deploying contracts with the account:', deployer.address);
  console.log('Account balance:', (await deployer.getBalance()).toString());

  // pass the registry instance to the deploy method
  const IExecWhitelist = await ethers.getContractFactory('IExecWhitelist');
  const iExecWhitelist = await IExecWhitelist.deploy();
  const receipt = await iExecWhitelist.deployed();
  
  console.log("Gas used: ",(await receipt.deployTransaction.wait()).gasUsed.toString())
  // save the smart contract address in `.smart-contract-address` file for next usages
  await saveSmartContractAddress(iExecWhitelist.address);
  
  console.log(
    'IExecWhitelist contract deployed to address:',
    iExecWhitelist.address
  );

};

main().catch((error) => {
  console.error(error);
  process.exitCode = 1;
});
```
Do the same thing with a more modern approach to the framework. 
Moreover i didn't see the interest to check if the smart contract is already deployed as many users can deploy the same one

File: test/IExecWhitelist.test.ts [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+      await expect(
+        whitelist
+          .connect(owner)
+          .remRessourceToWhitelist(await addr1.getAddress())
+      )
+        .to.emit(whitelist, 'KeyPurposeUpdate')
+        .withArgs(ethers.hexlify(paddedAddress), 4, false);
+    });
+  });
+});
```
Comment (Le-Caignec) [inline test/IExecWhitelist.test.ts]:
```suggestion
/* tslint:disable */
import { expect } from 'chai';
import { ethers } from 'hardhat';
import { loadFixture } from '@nomicfoundation/hardhat-network-helpers';

describe('IExecWhitelist', function () {
  async function deploySCFixture() {
    const [owner, addr1, addr2] = await ethers.getSigners();
    // Create an instance of DatasetRegistry using the address
    const IExecWhitelistFactory = await ethers.getContractFactory(
      'IExecWhitelist'
    );
    const iExecWhitelist = await IExecWhitelistFactory.deploy();
    await iExecWhitelist.deployed();

    return { iExecWhitelist, owner, addr1, addr2 };
  }

  describe('Whitelisting', () => {
    it('should allow the owner to add a dapp address to the iExecWhitelist', async () => {
      const { iExecWhitelist, owner, addr1 } = await loadFixture(deploySCFixture);
      await iExecWhitelist
        .connect(owner)
        .addRessourceToWhitelist(await addr1.getAddress());
      const paddedAddress =
        '0x' + '00'.repeat(12) + (await addr1.getAddress()).slice(2);
      expect(
        await iExecWhitelist.keyHasPurpose(ethers.utils.hexlify(paddedAddress), 4)
      ).to.equal(true);
    });

    it('should prevent non-owners from adding a dapp address to the iExecWhitelist', async () => {
      const { iExecWhitelist, addr2, addr1 } = await loadFixture(deploySCFixture);
      await expect(
        iExecWhitelist
          .connect(addr1)
          .addRessourceToWhitelist(await addr2.getAddress())
      ).to.be.revertedWith('Ownable: caller is not the owner');
    });

    it('should allow the owner to remove a dapp address from the iExecWhitelist', async () => {
      const { iExecWhitelist, owner, addr1 } = await loadFixture(deploySCFixture);
      await iExecWhitelist
        .connect(owner)
        .addRessourceToWhitelist(await addr1.getAddress());
      await iExecWhitelist
        .connect(owner)
        .remRessourceToWhitelist(await addr1.getAddress());

      const paddedAddress =
        '0x' + '00'.repeat(12) + (await addr1.getAddress()).slice(2);
      expect(
        await iExecWhitelist.keyHasPurpose(ethers.utils.hexlify(paddedAddress), 4)
      ).to.equal(false);
    });

    it('should prevent non-owners from removing a dapp address from the iExecWhitelist', async () => {
      const { iExecWhitelist, owner, addr1 } = await loadFixture(deploySCFixture);
      await iExecWhitelist
        .connect(owner)
        .addRessourceToWhitelist(await addr1.getAddress());
      await expect(
        iExecWhitelist
          .connect(addr1)
          .remRessourceToWhitelist(await addr1.getAddress())
      ).to.be.revertedWith('Ownable: caller is not the owner');
    });
  });

  describe('Events', function () {
    it('should emit KeyPurposeUpdate event when a dapp address is added', async () => {
      const { iExecWhitelist, owner, addr1 } = await loadFixture(deploySCFixture);
      const paddedAddress =
        '0x' + '00'.repeat(12) + (await addr1.getAddress()).slice(2);

      await expect(
        iExecWhitelist
          .connect(owner)
          .addRessourceToWhitelist(await addr1.getAddress())
      )
        .to.emit(iExecWhitelist, 'KeyPurposeUpdate')
        .withArgs(ethers.utils.hexlify(paddedAddress), 4, true);
    });

    it('should emit KeyPurposeUpdate event when a dapp address is removed', async () => {
      const { iExecWhitelist, owner, addr1 } = await loadFixture(deploySCFixture);
      const paddedAddress =
        '0x' + '00'.repeat(12) + (await addr1.getAddress()).slice(2);

      await iExecWhitelist
        .connect(owner)
        .addRessourceToWhitelist(await addr1.getAddress());
      await expect(
        iExecWhitelist
          .connect(owner)
          .remRessourceToWhitelist(await addr1.getAddress())
      )
        .to.emit(iExecWhitelist, 'KeyPurposeUpdate')
        .withArgs(ethers.utils.hexlify(paddedAddress), 4, false);
    });
  });
});
```
two changements : 
- use loadFixture which will reste the test environment each time
- use ethers from hardhat as it already export ethers

**Review by Le-Caignec — APPROVED** [review]
Can you make the correspondind documentation on confluence for the drone Ci as the other repo : [her](https://iexecproject.atlassian.net/wiki/spaces/IP/pages/3066757157/Whitelist+Smart+Contract)

**Comment by Le-Caignec** [comment]
add .smart-contract-address in.gitignore

File: package.json (hunk 2 above)
Comment (SeddikBellamine) [reply package.json]: actually only mocha is not used the others are used to by eslint and test script

File: hardhat.config.ts (hunk 4 above)
Comment (SeddikBellamine) [reply hardhat.config.ts]: useless to add import '@nomiclabs/hardhat-etherscan'; as it exists in hardhat toolbox

**Pull request merged**